### ISSUE TITLE
Bug---Custom loss with metrics [acc] when using sparse categories.

### ISSUE THREAD

**Issue by LCorleone** [keras-team/keras#10213]
when i use a custom loss and the second label is sparse.
if the model.compile like this:
model.compile(loss=['binary_crossentropy', self_loss], optimizer='SGD', loss_weights=[1, 1], metrics=['acc']).
the accuracy of the second output will not be correct, i dig into the source code [keras/engine/training] and find that:

for metric in metrics:
                if metric in ('accuracy', 'acc', 'crossentropy', 'ce'):
                    # custom handling of accuracy/crossentropy
                    # (because of class mode duality)
                    output_shape = K.int_shape(self.outputs[i])
                    if (output_shape[-1] == 1 or
                       self.loss_functions[i] == losses.binary_crossentropy):
                        # case: binary accuracy/crossentropy
                        if metric in ('accuracy', 'acc'):
                            metric_fn = metrics_module.binary_accuracy
                        elif metric in ('crossentropy', 'ce'):
                            metric_fn = metrics_module.binary_crossentropy
                    elif self.loss_functions[i] == losses.sparse_categorical_crossentropy:
                        # case: categorical accuracy/crossentropy
                        # with sparse targets
                        if metric in ('accuracy', 'acc'):
                            metric_fn = metrics_module.sparse_categorical_accuracy
                        elif metric in ('crossentropy', 'ce'):
                            metric_fn = metrics_module.sparse_categorical_crossentropy
                    else:
                        # case: categorical accuracy/crossentropy
                        if metric in ('accuracy', 'acc'):
                            metric_fn = metrics_module.categorical_accuracy
                        elif metric in ('crossentropy', 'ce'):
                            metric_fn = metrics_module.categorical_crossentropy

the source code will use categorical_crossentropy if you do not use sparse_categorical_crossentropy when compile. So, i think may be there is a bug. Hope for answering!
